devolutions-crypto-wayk/src/bastion/fle.rs

```diff
@@ -0,0 +1,203 @@
+use uuid::Uuid;
+
+pub const AAD_MAGIC_NUMBER: [u8; 4] = [0x53, 0x0A, 0x0F, 0x0E];
+pub const SAFE_PAYLOAD_PREFIX: &str = "|~Safe~|";
+pub const DEFAULT_BLOCK_SIZE: usize = 32;
+
+/// Encrypts field-level content using default padding size.
+///
+/// See `encrypt_field_with_block_size` for details.
+pub fn encrypt_field(master_key_id: Uuid, key: &[u8], content: &[u8]) -> Result<String, Error> {
+    encrypt_field_with_block_size(master_key_id, key, content, DEFAULT_BLOCK_SIZE)
+}
+
+/// Encrypts field-level content using provided padding size.
+///
+/// - Output is the encrypted content in Bastion safe Field Value Format (ie: |~Safe~|k1AjQZYciYvlPmJMnHjbb…)
+///
+/// # Interals
+///
+/// - AAD is composed of a 4-byte magic number, a 4-byte reserved field and the master key UUID in big-endian binary format.
+/// - A 24-byte initialization vector (IV) is randomly generated.
+/// - Content is padded using PKCS#7 padding mode.
+/// - Encryption using XChaCha2020-Poly1035.
+/// - Output is a buffer such as `[magic number (4) | reserved (4) | UUID (16) | IV (24) | Ciphertext (variable) | Tag (16)]`
+///     encoded in Bastion safe Field Value Format.
+pub fn encrypt_field_with_block_size(
+    master_key_id: Uuid,
+    key: &[u8],
+    content: &[u8],
+    block_size: usize,
+) -> Result<String, Error> {
+    use rand::rngs::OsRng;
+    use rand::Fill;
+
+    let pad_leftover = block_size - (content.len() % block_size);
+    let total_len = AAD_SIZE + IV_SIZE + content.len() + pad_leftover + TAG_SIZE;
+    let mut buffer = vec![0u8; total_len];
+
+    // associated data [magic number (4) | reserved (4) | UUID (16)]
+    let (aad, rest) = buffer.split_at_mut(AAD_SIZE);
+    {
+        let (aad_magic, aad_rest) = aad.split_at_mut(AAD_MAGIC_NUMBER_SIZE);
+        aad_magic.copy_from_slice(&AAD_MAGIC_NUMBER);
+        let (aad_reserved, aad_uuid) = aad_rest.split_at_mut(AAD_RESERVED_SIZE);
+        aad_reserved.copy_from_slice(&[0, 0, 0, 0]); // useless, but just to be extra safe…
+        aad_uuid.copy_from_slice(master_key_id.as_bytes());
+    }
+
+    // nonce / IV
+    let (nonce, rest) = rest.split_at_mut(IV_SIZE);
+    nonce.try_fill(&mut OsRng)?;
+    let nonce = XNonce::from_slice(nonce);
+
+    // actual data
+    let (data, rest) = rest.split_at_mut(rest.len() - TAG_SIZE);
+    data[..content.len()].copy_from_slice(&content);
+    Pkcs7::pad(data, content.len(), block_size).map_err(|_| Error::Padding)?;
+
+    // in-place encryption
+    let key = Key::from_slice(key);
+    let aead = XChaCha20Poly1305::new(key);
+    let tag = aead
+        .encrypt_in_place_detached(&nonce, aad, data)
+        .map_err(|_| Error::XChaCha20)?;
+
+    // finally, write tag
+    rest.copy_from_slice(&tag);
+
+    // our final payload (ie: |~Safe~|k1AjQZYciYvlPmJMnHjbb…)
+    let mut final_payload = String::from(SAFE_PAYLOAD_PREFIX);
+    base64::encode_config_buf(&buffer, base64::STANDARD, &mut final_payload);
+
+    Ok(final_payload)
```

Comment:
We might want to zeroize "buffer" at that point since it's already been copied by the base64 function. (There is a zeroize crate made for that)